### DISCUSSION TITLE
神經風格遷移 (Neural Style Transfer) 介紹

### DISCUSSION THREAD

**Discussion by lazywulf** [1101-GS4719-C/Python-Community-GS4719-C#1669] [Show and tell]
## 介紹
神經風格遷移 (下面簡稱 NST) 是一種優化技術：將兩張圖片 (內容圖與風格圖) 「融合」在一起，使生成出來的圖看起來畫的是內容圖的內容，卻是風格圖的風格。

聽起來很酷對吧？那這是如何做到的呢？它是利用卷積神經網路 (CNN) 學習一張圖片特徵的特性，將特徵提取出來並以此為依據優化輸出圖片。

### 卷積神經網路
卷積神經網絡 (CNN) 是一種在計算機科學和人工智慧領域中使用的技術。CNN就像是一位智能的照片分析師，他的工作是尋找照片中的特定元素。例如貓，他只要專注於尋找可能是貓的東西，像是耳朵、毛皮，或者貓的形狀。

其中，特徵圖是 CNN 透過計算得出的圖像，用來捕捉一張照片的特徵與模式。這張圖像的目的是讓我們更容易理解圖像中的信息。就像一位外行人可能不明白一張複雜的藝術作品，但如果我們將其拆解成一個個拼圖塊，他們可以更容易理解藝術家的意圖。神經風格遷移就是利用這點來去捕捉並轉移風格的。

## 如何進行的
- 警告：會被數學炸喔！
### 損失
損失是我們評估模型表現的指標，就像你可能會用分數來評價一個考試的結果一樣。那我們可以根據這個分數來「加強」(就是優化)。不過記得是要越少越好 (損失，顧名思義)。
我們可以把圖片當作一個矩陣 (矩陣數值可以決定顏色)。

#### 內容損失
![image](https://github.com/NCU-GS4719-Python/Python-Community/assets/103130674/3f84138f-62ae-4114-8bed-4a8ded8dc12c)
我們直接對內容圖與生成圖的每一格計算均方誤差。

#### 風格損失
![image](https://github.com/NCU-GS4719-Python/Python-Community/assets/103130674/25c9c057-3fd9-402f-a296-d563991be3d1)
對於風格圖跟生成圖計算格拉姆矩陣、標準化並計算均方差。格拉姆矩陣可以用來計算每張特徵圖間的關係 (很像共變異數矩陣)。

#### 總損失
![image](https://github.com/NCU-GS4719-Python/Python-Community/assets/103130674/44636d61-c09d-4aa7-9451-521a37e48eec)
上完權重後相加即是總損失囉！就依此做優化！

## 成果
![image](https://github.com/NCU-GS4719-Python/Python-Community/assets/103130674/2a5249d5-c102-46f7-8b70-74d1187b5ffc)
![image](https://github.com/NCU-GS4719-Python/Python-Community/assets/103130674/66f8c549-e8af-4926-9337-d61f2edcbdbe)
![image](https://github.com/NCU-GS4719-Python/Python-Community/assets/103130674/4f0e7826-7419-41b6-ae6e-624289a6924f)

有興趣可以逛逛本人的 github page 喔！實作程式碼在上面喔！
論文連結：https://arxiv.org/abs/1508.06576

**Comment by Fion-Lin** [comment]
挖很酷ㄟ! 請問他可以所以請問是會需要對兩張圖片都提取特徵後，再用卷積去優化嗎? 還是是只提取風格圖的特徵，在把它捲積到內容圖上呢?

**Reply by lazywulf** [reply]
只提取風格圖的特徵喔！內容圖做的動作 (損失函數的部分) 比較像是希望生出來的圖不要太離譜。風格圖的損失才是對他的特徵圖做事 (提取特徵)。
論文中實驗好像是從白噪音開始優化，不過直接從內容圖做起點會比較快。
